Source/CompanyCommunicator.Prep.Func/Export/Streams/DataStreamFacade.cs

```diff
@@ -66,35 +66,53 @@ public class DataStreamFacade : IDataStreamFacade
         /// <returns>the streams of user data.</returns>
         public async IAsyncEnumerable<IEnumerable<UserData>> GetUserDataStreamAsync(string notificationId)
         {
-            if (notificationId == null)
-            {
-                throw new ArgumentNullException(nameof(notificationId));
-            }
+            _ = notificationId ?? throw new ArgumentNullException(nameof(notificationId));
 
             var sentNotificationDataEntitiesStream = this.sentNotificationDataRepository.GetStreamsAsync(notificationId);
+            var isForbidden = false;
+
             await foreach (var sentNotifcations in sentNotificationDataEntitiesStream)
             {
-                List<User> userList = new List<User>();
+                var users = new List<User>();
+
+                // filter the recipient not found users.
+                var recipients = sentNotifcations.Where(sentNotifcation => !sentNotifcation.DeliveryStatus.Equals(SentNotificationDataEntity.RecipientNotFound, StringComparison.CurrentCultureIgnoreCase));
+
                 try
                 {
-                    // filter the recipient not found users.
-                    var users = await this.usersService.GetBatchByUserIds(
-                        sentNotifcations
-                        .Where(sentNotifcation => !sentNotifcation.DeliveryStatus.Equals(SentNotificationDataEntity.RecipientNotFound, StringComparison.CurrentCultureIgnoreCase))
-                        .Select(notitification => notitification.RowKey)
-                        .ToList()
-                        .AsGroups());
-                    userList = users.ToList();
+                    if (!isForbidden)
+                    {
+                        // Group the recipients as per the graph batch api.
+                        var groupRecipientsByAadId = recipients?
+                           .Select(notitification => notitification.RowKey)
+                           .ToList()
+                           .AsGroups();
+
+                        if (!groupRecipientsByAadId.IsNullOrEmpty())
+                        {
+                            users = (await this.usersService.GetBatchByUserIds(groupRecipientsByAadId))?.ToList();
+                        }
+                    }
                 }
                 catch (ServiceException serviceException)
                 {
                     if (serviceException.StatusCode != HttpStatusCode.Forbidden)
                     {
                         throw serviceException;
```

Comment:
throw;
propagate the exception, not re-throw

Reply:
done.